### PR TITLE
Md label font style options fork

### LINKED ISSUE #515: [Bug]: Custom Font style does not work if MDLabel is imported to namespace
### Description of the Bug

Even after following the docs in registering the custom font style, it does not work if MDLabel is imported.

### Not Working Code
```
KV = '''
Screen:
    MDLabel:
        text: "SfProLight"
        halign: "center"
        font_style: "SfProLight"
'''

from kivy.core.text import LabelBase
from kivy.lang import Builder
from kivymd.uix.boxlayout import MDBoxLayout
from kivy.clock import Clock
from kivymd.uix.label import MDLabel
from kivymd.app import MDApp
from kivymd.font_definitions import theme_font_styles



class MainApp(MDApp):
    def build(self):
        LabelBase.register(
            name="SfProLight",
            fn_regular='font_sfpro_light.ttf')

        theme_font_styles.append('SfProLight')
        print('theme font styles', theme_font_styles)
        self.theme_cls.font_styles["SfProLight"] = [
            "SfProLight",
            16,
            False,
            0.15
        ]
        return Builder.load_string(KV)


MainApp().run()
```

### Working Code:
```
KV = '''
Screen:
    MDLabel:
        text: "SfProLight"
        halign: "center"
        font_style: "SfProLight"
'''

from kivy.core.text import LabelBase
from kivy.lang import Builder
from kivymd.uix.boxlayout import MDBoxLayout
from kivy.clock import Clock

from kivymd.app import MDApp
from kivymd.font_definitions import theme_font_styles


class MainApp(MDApp):
    def build(self):
        LabelBase.register(
            name="SfProLight",
            fn_regular='font_sfpro_light.ttf')

        theme_font_styles.append('SfProLight')
        print('theme font styles', theme_font_styles)
        self.theme_cls.font_styles["SfProLight"] = [
            "SfProLight",
            16,
            False,
            0.15
        ]
        return Builder.load_string(KV)


MainApp().run()
```

### Error Log:
```
C:\Users\User\AppData\Local\Programs\Python\Python38\python.exe C:/Users/User/PycharmProjects/Bug.py
[INFO   ] [Logger      ] Record log in C:\Users\User\.kivy\logs\kivy_20-09-10_61.txt
[INFO   ] [deps        ] Successfully imported "kivy_deps.angle" 0.2.0
[INFO   ] [deps        ] Successfully imported "kivy_deps.glew" 0.2.0
[INFO   ] [deps        ] Successfully imported "kivy_deps.sdl2" 0.2.0
[INFO   ] [Kivy        ] v2.0.0rc3, git-20c14b2, 20200615
[INFO   ] [Kivy        ] Installed at "C:\Users\User\AppData\Local\Programs\Python\Python38\lib\site-packages\kivy\__init__.py"
[INFO   ] [Python      ] v3.8.3 (tags/v3.8.3:6f8c832, May 13 2020, 22:37:02) [MSC v.1924 64 bit (AMD64)]
[INFO   ] [Python      ] Interpreter at "C:\Users\User\AppData\Local\Programs\Python\Python38\python.exe"
[INFO   ] [Image       ] Providers: img_tex, img_dds, img_sdl2, img_pil (img_ffpyplayer, img_gif ignored)
[INFO   ] [Text        ] Provider: sdl2
[INFO   ] [Factory     ] 185 symbols loaded
[INFO   ] [KivyMD      ] v0.104.1 (installed at "C:\Users\User\AppData\Roaming\Python\Python38\site-packages\kivymd\__init__.py")
[INFO   ] [Window      ] Provider: sdl2
[INFO   ] [GL          ] Using the "OpenGL" graphics system
[INFO   ] [GL          ] GLEW initialization succeeded
[INFO   ] [GL          ] Backend used <glew>
[INFO   ] [GL          ] OpenGL version <b'4.6.0 NVIDIA 451.82'>
[INFO   ] [GL          ] OpenGL vendor <b'NVIDIA Corporation'>
[INFO   ] [GL          ] OpenGL renderer <b'GeForce RTX 2060/PCIe/SSE2'>
[INFO   ] [GL          ] OpenGL parsed version: 4, 6
[INFO   ] [GL          ] Shading version <b'4.60 NVIDIA'>
[INFO   ] [GL          ] Texture max size <32768>
[INFO   ] [GL          ] Texture max units <32>
[INFO   ] [Window      ] auto add sdl2 input provider
[INFO   ] [Window      ] virtual keyboard not allowed, single mode, not docked
[INFO   ] [GL          ] NPOT texture support is available
theme font styles ['H1', 'H2', 'H3', 'H4', 'H5', 'H6', 'Subtitle1', 'Subtitle2', 'Body1', 'Body2', 'Button', 'Caption', 'Overline', 'Icon', 'SfProLight']
 Traceback (most recent call last):
   File "C:\Users\User\AppData\Local\Programs\Python\Python38\lib\site-packages\kivy\lang\builder.py", line 705, in _apply_rule
     setattr(widget_set, key, value)
   File "kivy\weakproxy.pyx", line 35, in kivy.weakproxy.WeakProxy.__setattr__
   File "kivy\properties.pyx", line 498, in kivy.properties.Property.__set__
   File "kivy\properties.pyx", line 542, in kivy.properties.Property.set
   File "kivy\properties.pyx", line 533, in kivy.properties.Property.set
   File "kivy\properties.pyx", line 1253, in kivy.properties.OptionProperty.check
 ValueError: MDLabel.font_style is set to an invalid option 'SfProLight'. Must be one of: ['H1', 'H2', 'H3', 'H4', 'H5', 'H6', 'Subtitle1', 'Subtitle2', 'Body1', 'Body2', 'Button', 'Caption', 'Overline', 'Icon']
 
 During handling of the above exception, another exception occurred:
 
 Traceback (most recent call last):
   File "C:/Users/User/PycharmProjects/Bug.py", line 36, in <module>
     MainApp().run()
   File "C:\Users\User\AppData\Local\Programs\Python\Python38\lib\site-packages\kivy\app.py", line 949, in run
     self._run_prepare()
   File "C:\Users\User\AppData\Local\Programs\Python\Python38\lib\site-packages\kivy\app.py", line 919, in _run_prepare
     root = self.build()
   File "C:/Users/User/PycharmProjects/Bug.py", line 33, in build
     return Builder.load_string(KV)
   File "C:\Users\User\AppData\Local\Programs\Python\Python38\lib\site-packages\kivy\lang\builder.py", line 408, in load_string
     self._apply_rule(
   File "C:\Users\User\AppData\Local\Programs\Python\Python38\lib\site-packages\kivy\lang\builder.py", line 710, in _apply_rule
     raise BuilderException(rule.ctx, rule.line,
 kivy.lang.builder.BuilderException: Parser: File "<inline>", line 6:
 ...
       4:        text: "SfProLight"
       5:        halign: "center"
 >>    6:        font_style: "SfProLight"
 ...
 ValueError: MDLabel.font_style is set to an invalid option 'SfProLight'. Must be one of: ['H1', 'H2', 'H3', 'H4', 'H5', 'H6', 'Subtitle1', 'Subtitle2', 'Body1', 'Body2', 'Button', 'Caption', 'Overline', 'Icon']
   File "C:\Users\User\AppData\Local\Programs\Python\Python38\lib\site-packages\kivy\lang\builder.py", line 705, in _apply_rule
     setattr(widget_set, key, value)
   File "kivy\weakproxy.pyx", line 35, in kivy.weakproxy.WeakProxy.__setattr__
   File "kivy\properties.pyx", line 498, in kivy.properties.Property.__set__
   File "kivy\properties.pyx", line 542, in kivy.properties.Property.set
   File "kivy\properties.pyx", line 533, in kivy.properties.Property.set
   File "kivy\properties.pyx", line 1253, in kivy.properties.OptionProperty.check
 

Process finished with exit code 1
```

### PR DESCRIPTION
**Pull request by lemousehunter**
Fixes #515 : Bug where MDLabel.font_style does not update

### Description of Changes
Changed `OptionProperty` to `ObjectProperty`
Added `check_font_style()` to replace the validation function of `OptionProperty`
Code in question:

`font_style = ObjectProperty("Body1")` changed to `font_style = ObjectProperty("Body1")`
```:attr:`font_style` is an :class:`~kivy.properties.OptionProperty``` changed to ```:attr:`font_style` is an :class:`~kivy.properties.OptionProperty` ```
List of options is not changed as it is still applicable (they are in `self.theme_cls.font_style.keys()`)

Added:

- `self.check_font_styles() to .__init__()`

- ```def check_font_styles(self):
          if self.font_style not in list(self.theme_cls.font_styles.keys()):
              raise ValueError(
                  "MDLabel.font_style is set to an invalid option '{}'. "
                  "Must be one of: {}".format(
                      self.font_style, list(self.theme_cls.font_styles)
                  )
              )
```
@ArtemSBulgakov @HeaTTheatR

### CODE REVIEW

**Comment by podraco** [comment]
i must say i don't approve the changes, it might be a good idea, but it's incomplete and because of the nature of kivy, it will give you more trouble than to use a list property

What i believe you should use is a pointer to a dict instead of changing the property type itself. Is it possible to update the option list after the instance is created?

File: kivymd/uix/label.py
Comment (podraco) [inline kivymd/uix/label.py]: This will be only executed on python code, and avoided in kvlang

**Comment by lemousehunter** [comment]
@podraco I tried to use OptionProperty but that’s not possible as it is immutable. I also checked with regards to how it will affect the rest of the module but thus far I don’t see an issue- as it fulfills the validation function of the OptionProperty

File: kivymd/uix/label.py
Comment (podraco) [inline kivymd/uix/label.py]: This seems like a good idea, but it should be a bind function.

We don't know if the developer will be changing the MDLabel property on the run.

what i believe it must be done is to bind directly to theme_cls.font_styles, so every time it changes, all the MDlabel class updates it's `font_style`  **option** attribute

File: kivymd/uix/label.py
Comment (podraco) [inline kivymd/uix/label.py]: This line will only be executed in Python code, and be avoided completely in KVlang. You should move it to a call after `__init__` is done.

**Review by podraco — CHANGES_REQUESTED** [review]
Some change need to be done, it's a good approach but it will cause some miss behavior between python code and kvlang.

And further testing need to be done too.

**Comment by podraco** [comment]
> @podraco I tried to use OptionProperty but that’s not possible as it is immutable. I also checked with regards to how it will affect the rest of the module but thus far I don’t see an issue- as it fulfills the validation function of the OptionProperty

You're right, i just read the source and it's a cpython class and the option attr is not mutable since it's a copy instead of a pointer... (yes, i know there's no direct pointers to any object in python but usually, when you use a dict instance you have a "pointer" to the object's content)

In that case i see two possible solutions,
1. The option property could be a subclassed and override the troubling part to allow a dynamic update of the option attr (i'm not sure if it's possible, but you could try).
1. use an objectproperty that has a special bind to update everything manually.

Either way it's your call.

File: kivymd/uix/label.py
Comment (lemousehunter) [reply kivymd/uix/label.py]: Currently, this is a non-issue imo, as whenever `self.theme_cls.font_styles` updates, `self.font_style` will also accept whatever that `self.theme_cls.font_style` has, since it iterates over its dict keys, meaning if `self.theme_cls.font_style` is updated, `self.font_style` will not raise an error if it has been added to `self.theme_cls.font_style`.

However, I do propose adding a check fn in the `self.update_font_style()` fn such that instead of raising a `key_error` when an unknown font_style is set, it will raise `ValueError: MDLabel.font_style is set to an invalid option` etc. 

Have just updated this PR do take a look @podraco @ArtemSBulgakov @HeaTTheatR

File: kivymd/uix/label.py
Comment (lemousehunter) [reply kivymd/uix/label.py]: Oops I missed that out. Thanks for raising this up will use the `Clock.schedule_once` to schedule the `check_function` @podraco

**Comment by HeaTTheatR** [comment]
@lemousehunter 
```python
    font_style = ObjectProperty("Body1")                                       # <-------- invalid property
    """
    Label font style.
    Available options are: `'H1'`, `'H2'`, `'H3'`, `'H4'`, `'H5'`, `'H6'`,     # <-------- invalid description
    `'Subtitle1'`, `'Subtitle2'`, `'Body1'`, `'Body2'`, `'Button'`,
    `'Caption'`, `'Overline'`, `'Icon'`.
    :attr:`font_style` is an :class:`~kivy.properties.OptionProperty`.         # <-------- invalid property
    and defaults to `'Body1'`.
    """```

**Comment by HeaTTheatR** [comment]
@lemousehunter 
Снимок экрана 2020-09-12 в 20 09 24

**Comment by lemousehunter** [comment]
> 
> 
> @lemousehunter
> 
> ```python
>     font_style = ObjectProperty("Body1")                                       # <-------- invalid property
>     """
>     Label font style.
>     Available options are: `'H1'`, `'H2'`, `'H3'`, `'H4'`, `'H5'`, `'H6'`,     # <-------- invalid description
>     `'Subtitle1'`, `'Subtitle2'`, `'Body1'`, `'Body2'`, `'Button'`,
>     `'Caption'`, `'Overline'`, `'Icon'`.
>     :attr:`font_style` is an :class:`~kivy.properties.OptionProperty`.         # <-------- invalid property
>     and defaults to `'Body1'`.
>     """```
> ```

@HeaTTheatR Why is it an invalid property?? As you have shown in the screenshot, it says ObjectProperty
Also, the description is still valid as these options are still accepted by MDLabel with the new changes

**Comment by lemousehunter** [comment]
> 
> 
> @HeaTTheatR 
> Снимок экрана 2020-09-12 в 20 09 24

As can be seen in this screenshot, the property type in the descriptor has been changed to reflect the change in property type in the code.

**Comment by HeaTTheatR** [comment]
@lemousehunter 
```font_style = StringProperty("Body1")```

**Comment by lemousehunter** [comment]
> 
> 
> @HeaTTheatR 
> `font_style = StringProperty("Body1")`

Alright just updated the PR, do take a look at it :)

**Comment by HeaTTheatR** [comment]
@lemousehunter Okay, I'll run tests tomorrow.

File: kivymd/uix/label.py
Comment (podraco) [inline kivymd/uix/label.py]: A good practice is to use f strings; it's quicker for the python interpreter to use f strings, instead of calling format, i still don't know why. i believe it has something to do with the C implementation of it.
It also makes the text easier to read by the developer.

```py
raise ValueError(
                f"MDLabel.font_style is set to an invalid option '{self.font_style}'."
                f"Must be one of: {list(self.theme_cls.font_styles}")
```

File: kivymd/uix/label.py
Comment (podraco) [inline kivymd/uix/label.py]: i would change 
```py
def check_font_styles(self, dt):
```
to
```py
def check_font_styles(self, *dt):
```
It makes easier to call the function over the body of the class, and it means that `*dt` is not required but accepted (even if it's not used).

this means that 
```py
if self.check_font_styles("") is True:
```
should be changed to
```py
if self.check_font_styles() is True:
```
This avoids the creation of an object that contains "" inside, and be stored in `*dt`. this improves any app that uses multiple labels because all of them will run the same code. 

Makes easier the GC job and keeps memory usage as low as possible.

**Review by podraco — COMMENTED** [review]
This code looks better and for what i read, looks good. 

Sadly, I can't test it without compromising my pull request. 
i made changes to mdicon as well to fix some buttons issues.

File: kivymd/uix/label.py
Comment (lemousehunter) [reply kivymd/uix/label.py]: oooh thanks for the tip will amend my PR @podraco

File: kivymd/uix/label.py
Comment (lemousehunter) [reply kivymd/uix/label.py]: thats interesting :) will amend my PR @podraco

**Review by HeaTTheatR — COMMENTED** [review]
Your changes fail the tests in the following code example:

```python
KV = '''
Screen:
    MDLabel:
        text: "SfProLight"
        halign: "center"
        font_style: "SfProLight"
'''

from kivy.core.text import LabelBase
from kivy.lang import Builder
from kivymd.uix.boxlayout import MDBoxLayout
from kivy.clock import Clock
from kivymd.uix.label import MDLabel
from kivymd.app import MDApp
from kivymd.font_definitions import theme_font_styles



class MainApp(MDApp):
    def build(self):
        LabelBase.register(
            name="SfProLight",
            fn_regular='font_sfpro_light.ttf')

        theme_font_styles.append('SfProLight')
        print('theme font styles', theme_font_styles)
        self.theme_cls.font_styles["SfProLight"] = [
            "SfProLight",
            16,
            False,
            0.15
        ]
        return Builder.load_string(KV)


MainApp().run()
```

Output:

```python
 Traceback (most recent call last):
   File "/Users/macbookair/Opt/Venv-38/venv/lib/python3.8/site-packages/kivy/lang/builder.py", line 705, in _apply_rule
     setattr(widget_set, key, value)
   File "kivy/weakproxy.pyx", line 35, in kivy.weakproxy.WeakProxy.__setattr__
   File "kivy/properties.pyx", line 498, in kivy.properties.Property.__set__
   File "kivy/properties.pyx", line 545, in kivy.properties.Property.set
   File "kivy/properties.pyx", line 600, in kivy.properties.Property.dispatch
   File "kivy/_event.pyx", line 1248, in kivy._event.EventObservers.dispatch
   File "kivy/_event.pyx", line 1154, in kivy._event.EventObservers._dispatch
   File "/Users/macbookair/Projects/My-Project/KivyMD/kivymd/uix/label.py", line 118, in update_font_style
     if self.check_font_styles() is True:
   File "/Users/macbookair/Projects/My-Project/KivyMD/kivymd/uix/label.py", line 110, in check_font_styles
     raise ValueError(
 ValueError: MDLabel.font_style is set to an invalid option 'SfProLight'.Must be one of: ['H1', 'H2', 'H3', 'H4', 'H5', 'H6', 'Subtitle1', 'Subtitle2', 'Body1', 'Body2', 'Button', 'Caption', 'Overline', 'Icon', 'JetBrainsMono']
 
 During handling of the above exception, another exception occurred:
 
 Traceback (most recent call last):
   File "/Users/macbookair/Projects/My-Project/KivyMD/eeeee.py", line 34, in <module>
     MainApp().run()
   File "/Users/macbookair/Opt/Venv-38/venv/lib/python3.8/site-packages/kivy/app.py", line 949, in run
     self._run_prepare()
   File "/Users/macbookair/Opt/Venv-38/venv/lib/python3.8/site-packages/kivy/app.py", line 919, in _run_prepare
     root = self.build()
   File "/Users/macbookair/Projects/My-Project/KivyMD/eeeee.py", line 31, in build
     return Builder.load_string(KV)
   File "/Users/macbookair/Opt/Venv-38/venv/lib/python3.8/site-packages/kivy/lang/builder.py", line 408, in load_string
     self._apply_rule(
   File "/Users/macbookair/Opt/Venv-38/venv/lib/python3.8/site-packages/kivy/lang/builder.py", line 710, in _apply_rule
     raise BuilderException(rule.ctx, rule.line,
 kivy.lang.builder.BuilderException: Parser: File "<inline>", line 7:
 ...
       5:        text: "SfProLight"
       6:        halign: "center"
 >>    7:        font_style: "SfProLight"
 ...
 ValueError: MDLabel.font_style is set to an invalid option 'SfProLight'.Must be one of: ['H1', 'H2', 'H3', 'H4', 'H5', 'H6', 'Subtitle1', 'Subtitle2', 'Body1', 'Body2', 'Button', 'Caption', 'Overline', 'Icon', 'JetBrainsMono']
   File "/Users/macbookair/Opt/Venv-38/venv/lib/python3.8/site-packages/kivy/lang/builder.py", line 705, in _apply_rule
     setattr(widget_set, key, value)
   File "kivy/weakproxy.pyx", line 35, in kivy.weakproxy.WeakProxy.__setattr__
   File "kivy/properties.pyx", line 498, in kivy.properties.Property.__set__
   File "kivy/properties.pyx", line 545, in kivy.properties.Property.set
   File "kivy/properties.pyx", line 600, in kivy.properties.Property.dispatch
   File "kivy/_event.pyx", line 1248, in kivy._event.EventObservers.dispatch
   File "kivy/_event.pyx", line 1154, in kivy._event.EventObservers._dispatch
   File "/Users/macbookair/Projects/My-Project/KivyMD/kivymd/uix/label.py", line 118, in update_font_style
     if self.check_font_styles() is True:
   File "/Users/macbookair/Projects/My-Project/KivyMD/kivymd/uix/label.py", line 110, in check_font_styles
     raise ValueError(
 
```

**Comment by lemousehunter** [comment]
> ```python
> 'JetBrainsMono'
> ```

I have tested the above code example and it worked for me. In your log, it appears that you have registered 'JetBrainsMono' instead of "SfProLight", as can be seen here:

` ValueError: MDLabel.font_style is set to an invalid option 'SfProLight'.Must be one of: ['H1', 'H2', 'H3', 'H4', 'H5', 'H6', 'Subtitle1', 'Subtitle2', 'Body1', 'Body2', 'Button', 'Caption', 'Overline', 'Icon', 'JetBrainsMono']`
hence the error. @HeaTTheatR

**Comment by HeaTTheatR** [comment]
@lemousehunter Oh, sorry!

**Pull request merged**
